File: js/ai.js
// ==========================================
// 📁 6. ai.js
// ==========================================
async function createNewChat() {
    const newId = generateId();
    const newChat = {
        id: newId,
        title: 'New Chat',
        messages: []
    };
    state.chats.unshift(newChat);
    state.currentChatId = newId;
    await saveChatToDB(newChat);
    renderChatList();
    renderChatMessages();
    els.chatInput.focus();
}

async function switchChat(chatId) {
    state.currentChatId = chatId;
    renderChatList();
    renderChatMessages();
}

async function deleteChat(chatId, event) {
    if(event) event.stopPropagation();
    if(!confirm("Delete this chat history?")) return;

    state.chats = state.chats.filter(c => c.id !== chatId);
    await deleteChatFromDB(chatId);

    if (state.currentChatId === chatId) {
        if (state.chats.length > 0) {
            state.currentChatId = state.chats[0].id;
        } else {
            createNewChat();
            return; 
        }
    }
    renderChatList();
    renderChatMessages();
}

async function renameChat(chatId, currentTitle, event) {
    if(event) event.stopPropagation();
    
    const newTitle = await showPromptModal("Rename Chat", currentTitle);
    if (newTitle && newTitle.trim() !== "") {
        const chat = state.chats.find(c => c.id === chatId);
        if (chat) {
            chat.title = newTitle.trim();
            await saveChatToDB(chat);
            renderChatList();
        }
    }
}

function renderChatList() {
    els.chatList.innerHTML = '';
    state.chats.forEach(chat => {
        const div = document.createElement('div');
        div.className = `chat-list-item ${chat.id === state.currentChatId ? 'active' : ''}`;
        div.innerHTML = `
            <div class="truncate pr-6">${chat.title}</div>
            <div class="chat-item-actions">
                <div class="action-btn action-rename" onclick="renameChat('${chat.id}', '${escapeHtml(chat.title)}', event)" title="Rename"><i class="fa-solid fa-pen"></i></div>
                <div class="action-btn action-delete" onclick="deleteChat('${chat.id}', event)" title="Delete"><i class="fa-solid fa-trash"></i></div>
            </div>
        `;
        div.onclick = () => switchChat(chat.id);
        els.chatList.appendChild(div);
    });
}

function renderChatMessages() {
    els.chatHistory.innerHTML = '';
    const chat = state.chats.find(c => c.id === state.currentChatId);
    if (!chat) return;

    if (chat.messages.length === 0) {
        els.chatHistory.innerHTML = `
            <div class="text-center text-xs text-gray-400 mt-4">
                Ask a question about the loaded PDFs.
                <br>Uses <span class="font-mono">nomic-embed-text</span> & local LLM.
            </div>`;
        return;
    }

    chat.messages.forEach(msg => {
        let htmlContent = msg.html;

        if (msg.role === 'assistant' && msg.context && htmlContent.includes('jumpToCitation')) {
            const getTooltip = (chunk) => {
                const snippet = chunk.text.length > 150 ? chunk.text.substring(0, 150) + "..." : chunk.text;
                return `Source: ${chunk.docName} (Page ${chunk.pageNum})&#10;Text: ${snippet}`;
            };

            htmlContent = htmlContent.replace(
                /<span class="citation-chip" onclick="jumpToCitation\((\d+)\)">(\d+)<\/span>/g, 
                (match, idxStr, text) => {
                    const idx = parseInt(idxStr);
                    const chunk = msg.context[idx];
                    if (chunk) {
                        return `<span class="citation-chip" onclick="handleCitationClick(this)" data-doc="${chunk.docId}" data-page="${chunk.pageNum}" data-text="${encodeURIComponent(chunk.text)}" title="${escapeHtml(getTooltip(chunk))}">${text}</span>`;
                    }
                    return match;
                }
            );

            htmlContent = htmlContent.replace(
                /<span class="text-\[10px\] bg-blue-50 text-blue-600 border border-blue-200 px-2 py-1 rounded cursor-pointer hover:bg-blue-100" onclick="jumpToCitation\((\d+)\)">.*?<\/span>/g,
                (match, idxStr) => {
                    const idx = parseInt(idxStr);
                    const chunk = msg.context[idx];
                    if (chunk) {
                        return `<span class="text-[10px] bg-blue-50 text-blue-600 border border-blue-200 px-2 py-1 rounded cursor-pointer hover:bg-blue-100" onclick="handleCitationClick(this)" data-doc="${chunk.docId}" data-page="${chunk.pageNum}" data-text="${encodeURIComponent(chunk.text)}" title="${escapeHtml(getTooltip(chunk))}">[${idx+1}]</span>`;
                    }
                    return match;
                }
            );
        }
        appendMessageToDOM(msg.role, htmlContent);
    });
    els.chatHistory.scrollTop = els.chatHistory.scrollHeight;
}

function appendMessageToDOM(role, htmlContent) {
    const msgDiv = document.createElement('div');
    msgDiv.className = `ai-message ${role}`;
    msgDiv.innerHTML = `<div class="ai-bubble">${htmlContent}</div>`;
    els.chatHistory.appendChild(msgDiv);
    els.chatHistory.scrollTop = els.chatHistory.scrollHeight;
}

function cosineSimilarity(vecA, vecB) {
    let dotProduct = 0;
    let normA = 0;
    let normB = 0;
    for (let i = 0; i < vecA.length; i++) {
        dotProduct += vecA[i] * vecB[i];
        normA += vecA[i] * vecA[i];
        normB += vecB[i] * vecB[i];
    }
    if (normA === 0 || normB === 0) return 0;
    return dotProduct / (Math.sqrt(normA) * Math.sqrt(normB));
}

async function getEmbedding(text) {
    try {
        const response = await fetch('http://localhost:11434/api/embeddings', {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({ model: "nomic-embed-text:latest", prompt: text })
        });
        const data = await response.json();
        return data.embedding;
    } catch (error) {
        console.error("Embedding Error:", error);
        return null;
    }
}

async function generateLLMResponse(prompt, onChunk) {
    try {
        const payload = { 
            model: state.aiSettings.model, 
            prompt: prompt, 
            stream: !!onChunk,
            options: {
                temperature: state.aiSettings.temperature
            }
        };

        const response = await fetch('http://localhost:11434/api/generate', {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify(payload)
        });

        if (!onChunk) {
            const data = await response.json();
            return data.response;
        }

        const reader = response.body.getReader();
        const decoder = new TextDecoder("utf-8");
        let fullResponse = "";
        let buffer = "";

        while (true) {
            const { done, value } = await reader.read();
            if (done) break;
            
            buffer += decoder.decode(value, { stream: true });
            const lines = buffer.split('\n');
            
            buffer = lines.pop(); 
            
            for (const line of lines) {
                if (line.trim() === '') continue;
                try {
                    const parsed = JSON.parse(line);
                    if (parsed.response) {
                        fullResponse += parsed.response;
                        onChunk(fullResponse);
                    }
                } catch (e) {
                    console.error("Error parsing JSON line:", e, line);
                }
            }
        }

        if (buffer.trim() !== '') {
            try {
                const parsed = JSON.parse(buffer);
                if (parsed.response) {
                    fullResponse += parsed.response;
                    onChunk(fullResponse);
                }
            } catch (e) {
                console.error("Error parsing final JSON line:", e, buffer);
            }
        }

        return fullResponse;
    } catch (error) {
        console.error("LLM Error:", error);
        return "Error connecting to LLM. Please check your Ollama configuration and selected model.";
    }
}

async function indexDocuments(force = false) {
    if (state.isIndexing) return;
    
    const docIds = Object.keys(state.documents);
    const embeddedDocIds = new Set(state.embeddings.map(e => e.docId));
    const needsIndexing = force || docIds.some(id => !embeddedDocIds.has(id));

    if (!needsIndexing) {
        updateAIStatus(`Ready. Indexed ${state.embeddings.length} segments.`);
        return;
    }

    state.isIndexing = true;
    updateAIStatus("Indexing PDFs...");
    
    // Use dynamically configured chunk size
    const chunkSize = state.aiSettings.chunkSize || 500;
    const overlap = Math.floor(chunkSize * 0.1); // 10% overlap

    for (const docId of docIds) {
        if (embeddedDocIds.has(docId) && !force) continue; 

        const doc = state.documents[docId];
        const totalPages = Math.min(doc.pageCount, 50); 

        for (let i = 1; i <= totalPages; i++) {
            try {
                const page = await doc.pdfDoc.getPage(i);
                const textContent = await page.getTextContent();
                const pageText = textContent.items.map(item => item.str).join(' ').trim();
                
                if(pageText.length < 10) continue;

                for (let start = 0; start < pageText.length; start += (chunkSize - overlap)) {
                    const chunk = pageText.substring(start, start + chunkSize);
                    if (chunk.trim().length > 10) {
                        const vector = await getEmbedding(chunk);
                        if (vector) {
                            state.embeddings.push({
                                id: `chunk_${docId}_${i}_${start}`,
                                text: chunk,
                                vector: vector,
                                docId: docId,
                                docName: doc.name,
                                pageNum: i
                            });
                        }
                    }
                }
            } catch (e) { console.error("Error indexing page", i, e); }
        }
    }
    
    state.isIndexing = false;
    updateAIStatus(`Ready. Indexed ${state.embeddings.length} segments.`);
}

function updateAIStatus(msg) {
    els.aiStatus.innerText = msg;
}

// Utility function to extract clean text from HTML
function extractTextFromHTML(html) {
    const tempDiv = document.createElement('div');
    tempDiv.innerHTML = html;
    return tempDiv.innerText || tempDiv.textContent || "";
}

async function handleChat() {
    const question = els.chatInput.value.trim();
    if (!question) return;

    const chat = state.chats.find(c => c.id === state.currentChatId);
    if (!chat) return;

    if (chat.messages.length === 0) {
        chat.title = question.substring(0, 30) + (question.length > 30 ? "..." : "");
        saveChatToDB(chat);
        renderChatList();
    }

    // Process User Message
    chat.messages.push({ role: 'user', html: escapeHtml(question) });
    appendMessageToDOM('user', escapeHtml(question));
    
    els.chatInput.value = '';
    updateAIStatus("Thinking...");

    const questionVector = await getEmbedding(question);
    if (!questionVector) {
        appendMessageToDOM('assistant', "Failed to generate embedding for question.");
        updateAIStatus("Error");
        return;
    }

    // ---------------------------------------------------------
    // BUDGET RETRIEVAL STRATEGY (Using user configurations)
    // ---------------------------------------------------------
    const allScoredEmbeddings = state.embeddings
        .map(emb => ({ ...emb, score: cosineSimilarity(questionVector, emb.vector) }))
        .sort((a, b) => b.score - a.score);

    let scoredEmbeddings = [];
    let totalContextChars = 0;
    let highestScore = allScoredEmbeddings.length > 0 ? allScoredEmbeddings[0].score : 0;
    let lowestSelectedScore = highestScore;

    if (allScoredEmbeddings.length > 0) {
        if (allScoredEmbeddings[0].score < state.aiSettings.similarityThreshold) {
            // Fallback: If no chunk meets the threshold, use the single highest scoring chunk (let LLM reject it if Strict RAG is on)
            scoredEmbeddings.push(allScoredEmbeddings[0]);
            totalContextChars = allScoredEmbeddings[0].text.length;
            lowestSelectedScore = allScoredEmbeddings[0].score;
        } else {
            // Configurable Budget Retrieval
            for (const chunk of allScoredEmbeddings) {
                if (chunk.score < state.aiSettings.similarityThreshold) break; 
                if (scoredEmbeddings.length >= state.aiSettings.maxChunks) break; 
                if (totalContextChars + chunk.text.length > state.aiSettings.contextBudget) break;

                scoredEmbeddings.push(chunk);
                totalContextChars += chunk.text.length;
                lowestSelectedScore = chunk.score;
            }
        }
    }

    console.log(`[Retrieval] Selected Chunks: ${scoredEmbeddings.length}`);
    console.log(`[Retrieval] Total Characters: ${totalContextChars}`);
    console.log(`[Retrieval] Highest Score: ${highestScore.toFixed(4)}`);
    console.log(`[Retrieval] Lowest Selected Score: ${lowestSelectedScore.toFixed(4)}`);
    // ---------------------------------------------------------

    state.currentContextChunks = scoredEmbeddings;

    const contextText = scoredEmbeddings.map((emb, idx) => 
        `[${idx + 1}] Source: ${emb.docName} (Page ${emb.pageNum})\nText: ${emb.text}`
    ).join("\n---\n");

    // Construct Prompt via AI Settings 
    let systemPrompt = state.aiSettings.systemPrompt;
    
    if (state.aiSettings.strictRag) {
        systemPrompt += "\n\nSTRICT RAG INSTRUCTION: Answer ONLY using the provided Context. If the context does not contain the answer, reply exactly with: 'I don't know based on the provided context.' Do not use outside knowledge.";
    }

    const style = state.aiSettings.responseStyle;
    if (style === "Concise") {
        systemPrompt += "\n\nResponse Style Instruction: Be extremely concise and direct. Provide only the essential facts extracted from the context. Keep your response brief (2-3 sentences if possible) without unnecessary fluff or conversational filler.";
    } else if (style === "Expert") {
        systemPrompt += "\n\nResponse Style Instruction: Answer as a domain expert. Use precise, technical, and professional language. Provide a nuanced, highly rigorous analysis based on the context. Assume the reader possesses advanced technical knowledge.";
    } else {
        // Detailed (Default)
        systemPrompt += "\n\nResponse Style Instruction: Provide a thorough, comprehensive, and detailed explanation. Break down the information clearly, step-by-step. Use formatting like bullet points or bold text if helpful to make the detailed answer highly readable.";
    }

    // Chat History Construction
    let historyText = "";
    if (state.aiSettings.includeChatHistory && chat.messages.length > 1) {
        // Get the last 6 messages excluding the user's latest question (which is at the end)
        const recentMessages = chat.messages.slice(0, -1).slice(-6);
        if (recentMessages.length > 0) {
            historyText = "--- Recent Chat History ---\n";
            recentMessages.forEach(m => {
                const cleanText = extractTextFromHTML(m.html).trim();
                historyText += `${m.role === 'user' ? 'User' : 'Assistant'}: ${cleanText}\n\n`;
            });
            historyText += "---------------------------\n\n";
        }
    }

    const fullPrompt = `${systemPrompt}\n\nContext:\n${contextText}\n\n${historyText}User Question: ${question}\n\nAnswer:`;

    // Pre-build the sources footer HTML
    let sourcesHtml = '';
    if (scoredEmbeddings.length > 0) {
        sourcesHtml = `<div class="mt-2 pt-2 border-t border-gray-200 flex flex-wrap gap-1">`;
        scoredEmbeddings.forEach((emb, idx) => {
            const snippet = emb.text.length > 150 ? emb.text.substring(0, 150) + "..." : emb.text;
            const tooltipText = `Source: ${emb.docName} (Page ${emb.pageNum})&#10;Text: ${snippet}`;

            sourcesHtml += `<span class="text-[10px] bg-blue-50 text-blue-600 border border-blue-200 px-2 py-1 rounded cursor-pointer hover:bg-blue-100" onclick="handleCitationClick(this)" data-doc="${emb.docId}" data-page="${emb.pageNum}" data-text="${encodeURIComponent(emb.text)}" title="${escapeHtml(tooltipText)}">[${idx+1}]</span>`;
        });
        sourcesHtml += `</div>`;
    }

    if (state.aiSettings.skipLlm) {
        let finalHtml = "<em class='text-gray-500'>LLM generation skipped. Returning sources only.</em>";
        if (scoredEmbeddings.length === 0) {
            finalHtml = "<em class='text-gray-500'>LLM generation skipped. No relevant sources found for your query.</em>";
        }
        
        const assistantMsg = { role: 'assistant', html: `<div class="streaming-content">${finalHtml}</div>${sourcesHtml}`, context: scoredEmbeddings };
        chat.messages.push(assistantMsg);

        const msgDiv = document.createElement('div');
        msgDiv.className = `ai-message assistant`;
        msgDiv.innerHTML = `
            <div class="ai-bubble">
                <div class="streaming-content">${finalHtml}</div>
                ${sourcesHtml}
            </div>
        `;
        els.chatHistory.appendChild(msgDiv);
        els.chatHistory.scrollTop = els.chatHistory.scrollHeight;

        await saveChatToDB(chat);
        updateAIStatus("Ready");
        return;
    }

    const assistantMsg = { role: 'assistant', html: '', context: scoredEmbeddings };
    chat.messages.push(assistantMsg);

    const msgDiv = document.createElement('div');
    msgDiv.className = `ai-message assistant`;
    msgDiv.innerHTML = `
        <div class="ai-bubble">
            <div class="streaming-content"><span class="text-gray-400 italic">Generating response...</span></div>
            ${sourcesHtml}
        </div>
    `;
    els.chatHistory.appendChild(msgDiv);
    els.chatHistory.scrollTop = els.chatHistory.scrollHeight;
    
    const contentBubble = msgDiv.querySelector('.streaming-content');

    // Stream the LLM response
    const answer = await generateLLMResponse(fullPrompt, (fullText) => {
        let htmlText = fullText.replace(/\n/g, '<br>');
        
        htmlText = htmlText.replace(/\[(\d+)\]/g, (match, num) => {
            const idx = parseInt(num) - 1;
            const emb = scoredEmbeddings[idx];
            if (!emb) return match; 

            const snippet = emb.text.length > 150 ? emb.text.substring(0, 150) + "..." : emb.text;
            const tooltipText = `Source: ${emb.docName} (Page ${emb.pageNum})&#10;Text: ${snippet}`;

            return `<span class="citation-chip" onclick="handleCitationClick(this)" data-doc="${emb.docId}" data-page="${emb.pageNum}" data-text="${encodeURIComponent(emb.text)}" title="${escapeHtml(tooltipText)}">${num}</span>`;
        });
        
        contentBubble.innerHTML = htmlText;
        els.chatHistory.scrollTop = els.chatHistory.scrollHeight;
    });

    let finalHtml = answer.replace(/\n/g, '<br>');
    
    finalHtml = finalHtml.replace(/\[(\d+)\]/g, (match, num) => {
        const idx = parseInt(num) - 1;
        const emb = scoredEmbeddings[idx];
        if (!emb) return match; 

        const snippet = emb.text.length > 150 ? emb.text.substring(0, 150) + "..." : emb.text;
        const tooltipText = `Source: ${emb.docName} (Page ${emb.pageNum})&#10;Text: ${snippet}`;

        return `<span class="citation-chip" onclick="handleCitationClick(this)" data-doc="${emb.docId}" data-page="${emb.pageNum}" data-text="${encodeURIComponent(emb.text)}" title="${escapeHtml(tooltipText)}">${num}</span>`;
    });

    contentBubble.innerHTML = finalHtml;
    
    assistantMsg.html = `<div class="streaming-content">${finalHtml}</div>${sourcesHtml}`;
    
    els.chatHistory.scrollTop = els.chatHistory.scrollHeight;
    await saveChatToDB(chat);
    updateAIStatus("Ready");
}

window.handleCitationClick = function(el) {
    const docId = el.dataset.doc;
    const pageNum = parseInt(el.dataset.page);
    const text = decodeURIComponent(el.dataset.text);

    if (!docId || !state.documents[docId]) return;

    state.view.right.docId = docId;
    state.view.right.pageNum = pageNum;
    
    renderPage('right');

    setTimeout(() => {
        highlightChunk(text, 'right');
    }, 100);
};

window.jumpToCitation = function(index) {
    const chunk = state.currentContextChunks[index];
    if (!chunk) return;

    state.view.right.docId = chunk.docId;
    state.view.right.pageNum = chunk.pageNum;
    
    renderPage('right');

    setTimeout(() => {
        highlightChunk(chunk.text, 'right');
    }, 100);
};

File: pdf-linker-studio-server/server/routers/ai.py
"""
AI / RAG REST API (project-scoped).

All Ollama communication happens here on the server. The browser never calls
Ollama directly — it calls these endpoints instead.

Every endpoint requires the `X-Project-Id` header so AI/RAG indexes and chats
are isolated per project.
"""
import json
import re
from typing import Optional

from fastapi import APIRouter, HTTPException, Depends
from fastapi.responses import StreamingResponse
from pydantic import BaseModel

from .. import config
from .. import database as db
from ..deps import get_current_project
from ..services import ollama, embeddings

router = APIRouter()


class IndexRequest(BaseModel):
    force: bool = False


class SearchRequest(BaseModel):
    query: str
    top_k: int = 8
    min_score: float = 0.65
    context_budget: int = 4000


class ChatRequest(BaseModel):
    question: str
    system_prompt: str = "You are a helpful assistant answering questions based on the provided PDF context."
    chat_history: list = []
    model: Optional[str] = None
    temperature: float = 0.7
    response_style: str = "Detailed"
    strict_rag: bool = True
    skip_llm: bool = False
    include_chat_history: bool = True
    similarity_threshold: float = 0.65
    context_budget: int = 4000
    max_chunks: int = 8


@router.post("/index")
def index_endpoint(req: IndexRequest, project_id: str = Depends(get_current_project)):
    try:
        result = embeddings.index_documents(project_id, force=req.force)
        return result
    except Exception as e:
        raise HTTPException(500, f"Indexing failed: {e}")


@router.get("/index/status")
def index_status(project_id: str = Depends(get_current_project)):
    return embeddings.get_indexing_status(project_id)


@router.post("/search")
def search_endpoint(req: SearchRequest, project_id: str = Depends(get_current_project)):
    try:
        results = embeddings.search(
            project_id,
            req.query,
            top_k=req.top_k,
            min_score=req.min_score,
            context_budget=req.context_budget,
        )
        for r in results:
            if "error" in r:
                continue
            doc = db.query_one(
                "SELECT name FROM documents WHERE id = ? AND project_id = ?",
                (r.get("doc_id"), project_id)
            )
            r["docName"] = doc["name"] if doc else "Unknown"
        return {"results": results}
    except Exception as e:
        raise HTTPException(500, str(e))


@router.get("/status")
def status(project_id: str = Depends(get_current_project)):
    return {
        "ollama_reachable": ollama.check_connection(),
        "ollama_url": config.OLLAMA_URL,
        "embedding_model": config.OLLAMA_EMBEDDING_MODEL,
        "llm_model": config.OLLAMA_LLM_MODEL,
        "embedding_count": (db.query_one(
            "SELECT COUNT(*) as c FROM embeddings WHERE project_id = ?", (project_id,)
        ) or {}).get("c", 0),
    }


@router.post("/chat")
def chat_endpoint(req: ChatRequest, project_id: str = Depends(get_current_project)):
    """Streaming chat with retrieval (Server-Sent Events)."""
    def event_stream():
        try:
            search_results = embeddings.search(
                project_id,
                req.question,
                top_k=req.max_chunks,
                min_score=req.similarity_threshold,
                context_budget=req.context_budget,
            )
            if search_results and "error" in search_results[0]:
                yield f"data: {json.dumps({'type': 'error', 'message': search_results[0]['error']})}\n\n"
                return

            for r in search_results:
                doc = db.query_one(
                    "SELECT name, page_ids_json FROM documents WHERE id = ? AND project_id = ?",
                    (r.get("doc_id"), project_id)
                )
                r["docName"] = doc["name"] if doc else "Unknown"
                page_num = "?"
                if doc and doc.get("page_ids_json"):
                    try:
                        page_ids = json.loads(doc["page_ids_json"])
                        if r.get("page_id") in page_ids:
                            page_num = page_ids.index(r["page_id"]) + 1
                    except (ValueError, json.JSONDecodeError):
                        pass
                r["pageNum"] = page_num

            yield f"data: {json.dumps({'type': 'sources', 'data': search_results})}\n\n"

            if req.skip_llm:
                yield f"data: {json.dumps({'type': 'done', 'answer': '', 'skipped_llm': True})}\n\n"
                return

            context_text = "\n---\n".join(
                f"[{i+1}] Source: {r.get('docName')} (Page {r.get('pageNum')})\nText: {r.get('text', '')}"
                for i, r in enumerate(search_results)
            )
            history_text = ""
            if req.include_chat_history and req.chat_history:
                recent = req.chat_history[-6:]
                if recent:
                    history_text = "--- Recent Chat History ---\n"
                    for m in recent:
                        text = re.sub(r"<[^>]+>", "", m.get("html", "")).strip()
                        role = "User" if m.get("role") == "user" else "Assistant"
                        history_text += f"{role}: {text}\n\n"
                    history_text += "---------------------------\n\n"

            system_prompt = req.system_prompt
            if req.strict_rag:
                system_prompt += "\n\nSTRICT RAG INSTRUCTION: Answer ONLY using the provided Context. If the context does not contain the answer, reply exactly with: 'I don't know based on the provided context.' Do not use outside knowledge."

            if req.response_style == "Concise":
                system_prompt += "\n\nResponse Style Instruction: Be extremely concise and direct. Provide only the essential facts extracted from the context. Keep your response brief (2-3 sentences if possible) without unnecessary fluff or conversational filler."
            elif req.response_style == "Expert":
                system_prompt += "\n\nResponse Style Instruction: Answer as a domain expert. Use precise, technical, and professional language. Provide a nuanced, highly rigorous analysis based on the context. Assume the reader possesses advanced technical knowledge."
            else:
                system_prompt += "\n\nResponse Style Instruction: Provide a thorough, comprehensive, and detailed explanation. Break down the information clearly, step-by-step. Use formatting like bullet points or bold text if helpful to make the detailed answer highly readable."

            full_prompt = f"{system_prompt}\n\nContext:\n{context_text}\n\n{history_text}User Question: {req.question}\n\nAnswer:"

            full_answer = ""
            try:
                for token in ollama.generate_stream(full_prompt, model=req.model, temperature=req.temperature):
                    full_answer += token
                    yield f"data: {json.dumps({'type': 'token', 'text': token, 'full': full_answer})}\n\n"
            except Exception as e:
                yield f"data: {json.dumps({'type': 'error', 'message': f'LLM error: {e}'})}\n\n"
                return

            yield f"data: {json.dumps({'type': 'done', 'answer': full_answer, 'sources': search_results})}\n\n"
        except Exception as e:
            yield f"data: {json.dumps({'type': 'error', 'message': str(e)})}\n\n"

    return StreamingResponse(event_stream(), media_type="text/event-stream")


@router.get("/embeddings/{doc_id}")
def list_embeddings(doc_id: str, project_id: str = Depends(get_current_project)):
    rows = db.query_all(
        "SELECT id, page_id, text FROM embeddings WHERE doc_id = ? AND project_id = ?",
        (doc_id, project_id)
    )
    return {"chunks": [{"id": r["id"], "pageId": r["page_id"], "text": r["text"]} for r in rows]}
